Detecting duplicate resource and module names

miqm opened the pull request:
Adding detection of duplicate names for this same type of resource and modules in one file.

This functionality is to warn user that his deployment will fail validation process, as ARM does not allow multiple definitions of one resource to be declared in one deployment.

This same goes with modules, as they are compiled to nested deployments.

For modules and extension resources, the scope of the declaration must also match.

This detector only takes into account names that are plain text - it does not take into account names that uses interpolation - they will always be treated as different (for now).

Fixes #802

anthony-c-martin commented on src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs:
```diff
@@ -690,6 +691,18 @@ public ErrorDiagnostic ArgumentCountMismatch(int argumentCount, int mininumArgum
                 DiagnosticLevel.Error,
                 "BCP119",
                 $"Unsupported scope for extension resource deployment. Expected a resource reference.");
+
+            public Diagnostic ResourceMultipleDeclarations(string resourceNames) => new(
+                TextSpan,
+                DiagnosticLevel.Warning,
```
Warning will allow compilation to continue. Should this be `DiagnosticLevel.Error`?

miqm replied:
This is an open question - I think warning as properly built ARM JSON will be built (no compilation blocker) and problem will be on deployment (runtime)

anthony-c-martin replied:
We should always block compilation if we know we're going to generate something that is invalid. Warnings are just for scenarios where it's not possible to say whether or not the template is valid, or if we're trying to advise against a bad practice. So in this scenario, since we know for sure the template won't deploy, we should block it.

miqm replied:
Ok, will change to errors.

Although question on showing error itself - should I highlight only value (`'resource-name'`) that is invalid or entire property (`name: 'resource-name'`)?

alex-frankel replied:
I think only the value, as there is nothing wrong with the property called `name`, just the value of it. Same as we do for other property values:

![image](https://user-images.githubusercontent.com/11078275/103418943-84aaf880-4b56-11eb-902b-b82c3e90a192.png)

miqm replied:
done.

anthony-c-martin commented on src/Bicep.Core/Emit/EmitLimitationCalculator.cs:
```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+
+            foreach (var duplicatedResourceGroup in duplicateResources)
+            {
+                var duplicatedResourceNames = string.Join(", ", duplicatedResourceGroup.Select(x => x.ResourceName));
```
Please move list formatting logic into the DiagnosticBuilder method (see the use of `ToQuotedString` [here](https://github.com/Azure/bicep/blob/19140c30917d0129f84969f9070dd97110aded6b/src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs#L211) for an example)

miqm replied:
moved.

anthony-c-martin commented on src/Bicep.Core/Emit/EmitLimitationCalculator.cs:
```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+            // This method only checks, if in one deployment we do not have 2 or more resources with this same name in one deployment to avoid template validation error
+            // This will not check resource constraints such as necessity of having unique virtual network names within resource group
+
+            var duplicateResources = semanticModel.Root.AllDeclarations
+                .OfType<ResourceSymbol>()
```
Use `semanticModel.Root.ResourceDeclarations` to access resource symbols

miqm replied:
done.

anthony-c-martin commented on src/Bicep.Core/Emit/EmitLimitationCalculator.cs:
```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+
+            var duplicateResources = semanticModel.Root.AllDeclarations
+                .OfType<ResourceSymbol>()
+                .Where(res => res.Type is ResourceType && res.DeclaringResource.Body is ObjectSyntax)
+                .Select(res =>
+                {
+                    var namePropSyntax = ((ObjectSyntax)res.DeclaringResource.Body).SafeGetPropertyByName(LanguageConstants.ResourceNamePropertyName);
+                    var resourceScope = resourceScopeData[res];
+                    return new
+                    {
+                        ResourceName = res.Name,
+                        ResourceScope = resourceScope,
+                        ResourceTypeFQDN = ((ResourceType)res.Type).TypeReference.FullyQualifiedType,
+                        ResourceNamePropertyValue = namePropSyntax?.Value as StringSyntax, //TODO: currently limiting check to property values that are strings, although it can be references or other syntaxes
+                        ResourceNamePropertySyntax = namePropSyntax
+                    };
+                })
+                .Where(x => x.ResourceNamePropertySyntax is not null && x.ResourceNamePropertyValue is not null)
```
Typing this without an IDE so some of the syntax may be incorrect, but I think it would be worth:
1. Create a dedicated class for this rather than using anonymous objects, and encapsulating the equality comparer, e.g.:
    ```csharp
    private class ResourceDefinition
    {
        public ResourceSymbol Resource { get; }
        public SyntaxBase NameSyntax { get; }
        public ResourceType Type { get; }
        public ResourceSymbol? Scope { get; }

        public ResourceDefinition(ResourceSymbol resource, SyntaxBase nameSyntax, ResourceType type, ResourceSymbol? scope)
        {
            Resource = resource;
            NameSyntax = nameSyntax;
            Type = type;
            Scope = scope;
        }

        public static IEqualityComparer<ResourceDefinition> Comparer { get; } = new EqualityComparer();

        private class EqualityComparer : IEqualityComparer<ResourceDefinition>
        {
            ...
        }
    }
    ```
1. Create a generator func to return valid `ResourceDefinition` instances for a given semantic model to avoid the complex Linq expressions:
    ```csharp
    public static IEnumerable<ResourceDefinition> GetResourceDefinitions(SemanticModel model, ImmutableDictionary<ResourceSymbol, ResourceSymbol?> resourceScopeData)
    {
        foreach (var resource in model.Root.ResourceDeclarations)
        {
            var nameSyntax = resource.SafeGetBodyPropertyValue(LanguageConstants.ResourceNamePropertyName);
            var resourceType = resource.Type as ResourceType;
            if (nameSyntax is null or resourceType is null)
            {
                continue;
            }
            
            yield return new ResourceDefinition(resource, nameSyntax, resourceType, resourceScopeData[resource]);
        }
    }
    ```
1. Above logic can then be simplified to:
    ```csharp
    var duplicatedResources = GetResourceDefinitions(semanticModel)
        .GroupBy(ResourceDefinition.Comparer)
        .Where(group => group.Count() > 1);

    foreach (var duplicatedResource in duplicateResources)
    {
        ...
    }
    ```

miqm replied:
Refactored

anthony-c-martin commented on src/Bicep.Core/Emit/EmitLimitationCalculator.cs:
```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+                var duplicatedModuleNames = string.Join(", ", duplicatedModuleGroup.Select(x => x.ModuleName));
+                foreach (var duplicatedModule in duplicatedModuleGroup)
+                {
+                    yield return DiagnosticBuilder.ForPosition(duplicatedModule.ModulePropertyNameSyntax!).ModuleMultipleDeclarations(duplicatedModuleNames);
```
Informally we're trying to keep `!` out of the (non-test) code, as it's a potential place for bugs to appear in the future. Unfortunately C# is often not super smart at tracking nullability, but there's usually a way to rewrite so that the compiler figures it out (e.g. rewriting so that the `!` is not necessary). Is that possible here?

miqm replied:
Will try to avoid it, perhaps with the dedicated class I'll be able to avoid using `!` operator

miqm replied:
avoided :)

anthony-c-martin commented on src/Bicep.Core/Emit/EmitLimitationCalculator.cs:
```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+
+            public int GetHashCode((string ResourceTypeFQDN, ResourceSymbol? ResourceScope,  StringSyntax PropertyNameValue) obj)
+            {
+                return HashCode.Combine(obj.ResourceTypeFQDN.ToLowerInvariant(), obj.ResourceScope, obj.PropertyNameValue.TryGetLiteralValue()?.ToLowerInvariant());
```
I've never benchmarked this, but I'd assume that using `StringComparer.OrdinalIgnoreCase.GetHashCode(<string>)` would be more performant than `<string>.ToLowerInvariant()` as it avoids allocation - e.g.:

```csharp
return HashCode.Combine(
    StringComparer.OrdinalIgnoreCase.GetHashCode(obj.ResourceTypeFQDN),
    obj.ResourceScope,
    StringComparer.OrdinalIgnoreCase.GetHashCode(obj.PropertyNameValue.TryGetLiteralValue() ?? ""));

miqm replied:
That might be true, didn't check that either. Of course in a result we will be building hash from values' hashes but that should be fine, as this is only for initial equality detection. Will change that.

anthony-c-martin replied:
> building hash from values' hashes

Yeah I've actually seen very little info on how `HashCode.Combine()` works, but I'd hope that it's optimized for handling int32 values - as `[int].GetHashCode()` is effectively a no-op.

Usually I do something like `GetHashCode(..) => val1.GetHashCode() ^ val2.GetHashCode() ^ val3.GetHashCode();`, but I suspect `HashCode.Combine()` is probably optimized to give a better distribution of hash values than using XOR.

miqm replied:
I used `new HashCode()` to build hash codes as described https://docs.microsoft.com/en-us/dotnet/api/system.hashcode?view=dotnet-plat-ext-5.0#instance-methods

miqm commented:
@anthony-c-martin I refactored PR as you suggested, looks way better now :) I've converted it to draft temporary, I'd like to add Unit tests for Resource/Module definitions comparers.

Although I've a question here - I placed classes in Bicep.Core.Utils namespace, however in UnitTests projects, Utils namespace contains test utils (utilities to run tests) and tests for Core Utils (utilities for core).

Should I place test cases there or separate it to some other namespace?

anthony-c-martin commented on bcode.ps1:
```diff
@@ -0,0 +1,8 @@
+# VSCode extension itself does not change that much and majority of changes are done in Language Server or Bicep.Core
+# This script allows you to quickly launch your existing VSCode with Bicep Extension installed but with your debug build of Bicep.LangServer and Bicep.CLI debug output added o the Path
+# any agruments passed to this script is given to the VSCode executable, therefore you can run
+# ./bcode.ps1 docs/Examples to open VSCode in the Examples folder running your local bicep lang server and in vscode terminal you have access to your local bicep CLI build
+
+$ENV:BICEP_LANGUAGE_SERVER_PATH=[System.IO.Path]::Join($PSScriptRoot,".\src\Bicep.LangServer\bin\Debug\net5.0\Bicep.LangServer.dll")
+$ENV:Path=[System.IO.Path]::Join($PSScriptRoot,".\src\Bicep.Cli\bin\Debug\net5.0") + ";" + $ENV:Path
+& Start-Process -FilePath "code" -ArgumentList $args
```
Please could you create a directory `/scripts` and move this file there?

miqm replied:
moved.

anthony-c-martin reviewed (approved):
Looks great, thank you for doing this!

majastrz commented:
Indeed thanks for getting it done. However, the PR is still in draft state and can't be merged.

miqm commented:
yes, there're 2 things I still have to do: unit test for comparers and moving the script. Will do by the end of the week.

miqm commented:
@anthony-c-martin @majastrz - I moved the script into `scripts` folder as requested. As for Unit Testing, Comparers rely heavily on `StringSyntax` and `ResourceSymbol` classes that cannot be easily mocked, therefore I think we need to rely on integration tests for them.

miqm commented:
@alex-frankel, @anthony-c-martin  - can we merge it?

TarunSunkaraneni commented:
Thanks @miqm!
